lib/geckodriver.js

```diff
@@ -0,0 +1,152 @@
+// transpile:main
+import events from 'events';
+import _ from 'lodash';
+import { logger } from 'appium-support';
+import { JWProxy } from 'appium-base-driver';
+import { SubProcess } from 'teen_process';
+
+const log = logger.getLogger('Geckodriver');
+
+const DEFAULT_HOST = '127.0.0.1';
+const DEFAULT_PORT = 29146;
+const SYSTEM_GECKODRIVER = '/usr/local/bin/geckodriver';
+const DEFAULT_INIT_URL = 'https://www.mozilla.org/';
+
+class Geckodriver extends events.EventEmitter {
+  constructor (args = {}) {
+    super();
+
+    const {
+      host = DEFAULT_HOST,
+      port = DEFAULT_PORT
+    } = args;
+
+    this.proxyHost = host;
+    this.proxyPort = port;
+    this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort, base: ''});
+    this.state = Geckodriver.STATE_STOPPED;
+    this.capabilities = {};
+  }
+
+  async start (caps, emitStartingState = true) {
+    this.capabilities = _.cloneDeep(caps);
+
+    const args = [`--host=${this.proxyHost}`, `--port=${this.proxyPort}`, '-vv'];
+
+    const startDetector = (stdout) => {
+      return stdout.indexOf('Listening') !== -1;
+    };
+
+    if (emitStartingState) {
+      this.changeState(Geckodriver.STATE_STARTING);
+    }
+
+    let processIsAlive = false;
+    try {
+      this.geckodriver = SYSTEM_GECKODRIVER;
+      this.proc = new SubProcess(this.geckodriver, args);
+      processIsAlive = true;
+
+      this.proc.on('output', (stdout, stderr) => {
+        if (this.verbose) {
+          for (let line of (stdout || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.debug(`[STDOUT] ${line}`);
+          }
+          for (let line of (stderr || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.error(`[STDERR] ${line}`);
+          }
+        }
+      });
+
+      this.proc.on('exit', (code, signal) => {
+        log.info(code, signal);
+        processIsAlive = false;
+
+        if (this.state !== Geckodriver.STATE_STOPPED &&
+          this.state !== Geckodriver.STATE_STOPPING &&
+          this.state !== Geckodriver.STATE_RESTARTING) {
+          let msg = `Geckodriver exited unexpectedly with code ${code}, ` +
+                  `signal ${signal}`;
+          log.error(msg);
+          this.changeState(Geckodriver.STATE_STOPPED);
+        }
+      });
+
+      log.info(`Spawning geckodriver with: ${this.geckodriver} ` +
+              `${args.join(' ')}. Only geckodriver 0.26.0 supports Android testing`);
```

Comment:
How about calling `geckodriver --help` and get the version from the stdout to then fail the process if the version is below `0.26.0`

Reply:
I was trying to avoid an extra call to geckodriver but implemented, adding _.memoize